### PR TITLE
zulip: Add authentication before initializing client.

### PR DESCRIPTION
**Pull request by TAM360**
Before sending any message to a user/stream via CLI, user's credentials are verified first. The idea is to update the users before they start using it (#533). This is done using **validate_credentials()** method which returns returns the result of **get_profile** API. The status of result is used to determine if the credentials are valid or not.

### CODE REVIEW

File: zulip/zulip/__init__.py [hunk 1]
```diff
@@ -331,7 +341,7 @@ def __init__(self, email: Optional[str] = None, api_key: Optional[str] = None, c
             if cert_bundle is None and config.has_option("api", "cert_bundle"):
                 cert_bundle = config.get("api", "cert_bundle")
             if insecure is None and config.has_option("api", "insecure"):
-                # Be quite strict about what is accepted so that users don't
+                # Be quite strict about what is ainvalidccepted so that users don't
```
Comment (aero31aero) [inline zulip/zulip/__init__.py:344]: Accidental change?

File: zulip/zulip/send.py [hunk 2]
```diff
@@ -71,28 +71,32 @@ def main() -> int:
...
+                'content': options.message,
+                'to': options.recipients,
+            }
+
+        if not do_send_message(client, message_data):
+            return 1
+        return 0
+    except zulip.InvalidCredentialsError as error:
+        logging.error(error)
+        return 0
```
Comment (aero31aero) [inline zulip/zulip/send.py:99]: We should return a non-zero value here. 0 implies the process went smoothly, without errors.

File: zulip/zulip/__init__.py [hunk 3]
```diff
@@ -226,12 +226,18 @@ def init_from_options(options: Any, client: Optional[str] = None) -> 'Client':
...
-                  client_cert_key=options.client_cert_key)
+
+    new_client = Client(email=options.zulip_email, api_key=options.zulip_api_key,
+                        config_file=options.zulip_config_file, verbose=options.verbose,
+                        site=options.zulip_site, client=client,
+                        cert_bundle=options.cert_bundle, insecure=options.insecure,
+                        client_cert=options.client_cert,
+                        client_cert_key=options.client_cert_key)
+
+    if new_client.get_profile()["result"] == "error":
```
Comment (aero31aero) [inline zulip/zulip/__init__.py:237]: Awesome, this sounds like the code we want. Just one thing I'd change here: we have some more places that also call `Client()` to create a new client. (`git grep 'Client(' | wc -l` shows 31 cases, most of which are calls to `zulip.Client()`. We also want to throw this error in thjose cases, thus, we should have this block of code in the Client class, just before it returns the client object.

File: zulip/zulip/__init__.py (hunk 1 above)
Comment (TAM360) [reply zulip/zulip/__init__.py:344]: Yup

File: zulip/zulip/__init__.py (hunk 3 above)
Comment (TAM360) [reply zulip/zulip/__init__.py:237]: I thought 0 meant an error. does -1 work in this case?

File: zulip/zulip/send.py [hunk 4]
```diff
@@ -71,28 +71,32 @@ def main() -> int:
...
+            message_data = {
+                'type': 'private',
+                'content': options.message,
+                'to': options.recipients,
+            }
+
+        if not do_send_message(client, message_data):
+            return 1
+        return 0
+    except zulip.InvalidCredentialsError as error:
```
Comment (neiljp) [inline zulip/zulip/send.py:97]: It might be clearer if we moved the try block to just be around the relevant code that raises the exception?

Comment (TAM360) [reply zulip/zulip/send.py:97]: I think it makes sense to keep the `client = zulip.init_from_options(options)` part in try/except block and move others outside.

**Comment by showell** [comment]
@TAM360 Do you think you have time in the next couple days to address the remaining comments?  It looks like it's mostly fixing a typo and limiting the scope of the try/except block.

**Comment by TAM360** [comment]
@showell yeah I do. I am gonna try to finish this by Sunday.

**Comment by showell** [comment]
@TAM360 Can you rebase these commits into a single commit?  We don't really want or need the merge commits or the commits that are just code-review fixes.

**Comment by TAM360** [comment]
I've already pushed 3-4 small commits to my branch. Will they get affected if I rebase my commits now and then push it?

**Comment by showell** [comment]
> 
> 
> @showell I've already pushed 3-4 small commits to my branch. Will they get affected if I rebase my commits now and then push it?

Yes, but that's the idea behind rebasing--I **want** you to replace those.

**Comment by TAM360** [comment]
@showell all right.

**Comment by aero31aero** [comment]
From my previous comment:

> We should have this block of code in the Client class, just before it returns the client object.

It looks like you've added the code to throw the error in multiple places; we only need it in the zulip Client's constructor.

**Comment by TAM360** [comment]
@aero31aero I've added the validation logic in Client's constructor but as a result it's failing 4 of the unit test. Should I update those tests or add new ones?

**Comment by TAM360** [comment]
@zulipbot abandon

**Comment by andersk** [comment]
(I’m not sure this PR is still active, but it’s apparently still being referenced.)

`"result": "error"` doesn’t necessarily imply that the credentials are wrong. We also need to check for HTTP status 401 or `"code": "INVALID_API_KEY"`.

Also, `sys.exit()` returns a successful exit code that’s inappropriate for an error condition. I think in most cases we just want to let the exception propagate up.

**Comment by showell** [comment]
I am gonna close this since @TAM360 has abandoned it.  @TAM360 Thanks for working on this!

**Pull request closed without merging**